src/machine.ts

```diff
@@ -0,0 +1,66 @@
+import { Path, PathType } from './path';
+
+export class State {
+  x: number;
+  y: number;
+  z: number;
+  e: number;
+  tool: number;
+  units: 'mm' | 'in';
+
+  static get initial(): State {
+    const state = new State();
+    Object.assign(state, { x: 0, y: 0, z: 0, e: 0, tool: 0, units: 'mm' });
+    return state;
+  }
+}
+
+export class Machine {
+  paths: Path[];
+  state: State;
+
+  constructor(state?: State) {
+    this.paths = [];
+    this.state = state || State.initial;
+  }
+
+  isPlanar(): boolean {
+    return (
+      this.paths.find(
+        (path) =>
+          path.travelType === PathType.Extrusion && path.vertices.some((_, i, arr) => i % 3 === 2 && arr[i] !== arr[2])
+      ) === undefined
+    );
+  }
+
+  layers(): Path[][] | null {
+    if (!this.isPlanar()) {
+      return null;
+    }
```

Comment:
I kinda like grouping layers together only when they are actual layers. 

Right now, it forces me to decouple the rendering from the layers. It breaks some of the nice features we have, but they most likely can be implemented differently.

Reply:
I'm aiming for v3, but breaking changes still need a good justification. Layers are very useful and lib users have a good chance of using them.

Reply:
As discussed in Discord, this PR is directly for v3.x and will merge in the `v3.x` branch. 

However, there should at least have an implementation using the tolerance logic that's in the current version. Will follow-up with a PR addressing that.